src/protections/MentionFlood.ts

```diff
@@ -0,0 +1,153 @@
+const PORT_REGEX = "(:[0-9]+)?";
+
+
+export class MentionFlood extends Protection {
+
+    settings = {
+        minutesBeforeTrusting: new NumberProtectionSetting(DEFAULT_MINUTES_BEFORE_TRUSTING),
+        maxMentionsPerMessage: new NumberProtectionSetting(DEFAULT_MAX_MENTIONS_PER_MESSAGE),
+        redact: new BooleanProtectionSetting(DEFAULT_REDACT),
+        action: new StringProtectionSetting()
+    };
+
+    private justJoined: Map<string, Map<string, Date>> = new Map<string, Map<string, Date>>();
+    private mention: RegExp;
+
+    constructor() {
+        super();
+        this.mention = new RegExp(`@${LOCALPART_REGEX}:(${DOMAIN_REGEX}|${IPV4_REGEX}|${IPV6_REGEX})${PORT_REGEX}`, "gi");
+    }
+
+    public get name(): string {
+        return 'MentionFlood';
+    }
+
+    public get description(): string {
+        return "If a user posts more mentions than a set amount of time after joining, they " +
+            "will be banned from that room.  This will not publish the ban to a ban list.";
+    }
+
+    public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any): Promise<any> {
+
+        const content = event['content'] || {};
+        const minsBeforeTrusting = this.settings.minutesBeforeTrusting.value;
+
+        if (minsBeforeTrusting > 0) {
+            if (!this.justJoined.get(roomId)) this.justJoined.set(roomId, new Map<string, Date>());
+
+            // When a new member logs in, store the time they joined.  This will be useful
+            // when we need to check if a message was sent within 20 minutes of joining
+            if (event['type'] === 'm.room.member') {
+                if (isTrueJoinEvent(event)) {
+                    const now = new Date();
+                    this.justJoined.get(roomId)?.set(event['state_key'], now);
+                    LogService.info("MentionFlood", `${htmlEscape(event['state_key'])} joined ${roomId} at ${now.toDateString()}`);
+                } else if (content['membership'] === 'leave' || content['membership'] === 'ban') {
+                    this.justJoined.get(roomId)?.delete(event['sender']);
+                }
+
+                return;
+            }
+        }
+
+        if (event['type'] === 'm.room.message') {
+            const message: string = content['formatted_body'] || content['body'] || "";
```

Comment:
do any clients opt for showing `body` even if `formatted_body` is present? if so we'll want to check both

Reply:
Clients show formatted_body if preferable, but mentions (currently) are only checked per the body, the formatted body does not get inspected for mentions/pings

Reply:
The order chosen here is mostly as only the formatted body of ele Web includes the mxid. The body only has the display names. And the code explicitly checks for mxids

Reply:
that makes sense.

Reply:
this is definitely the case for pill-style mentions (I've just checked) but otherwise I don't think so. this will probably be good enough for now though.

Reply:
The alternative would be just to check both and go with the higher number I guess